### PR TITLE
Unified symbolic shape variables between AOTAutograd and Inductor

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86659

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -329,28 +329,62 @@ def _get_key(self):
...
+        return [self.create_symintnode(i) for i in size], [self.create_symintnode(i) for i in stride]
+
+    def create_symintnode(self, expr: Union["sympy.Expr", int]):
+        if isinstance(expr, int):
+            return expr
+        py_sym_int = PySymInt(expr, self)
+        cpp_sym_int = torch.SymIntNode.new_symint(py_sym_int)  # type: ignore[attr-defined]
+        return cpp_sym_int
+
+    def __getitem__(self, val: int) -> "sympy.Expr":
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:372]: Nit, I'd probably prefer this not to be a magic method and just be a plain method. But maybe you need this for compat with inductor?

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -329,28 +329,62 @@ def _get_key(self):
...
-    # sizes are allocated via a different mechanism
-    def create_symint(self, name, val):
-        assert val >= 0
+    def create_symbolic_sizes_strides(self, ex: torch.Tensor):
+        size = [self[i] for i in ex.size()]
+        stride = [None] * len(size)
+        for i, val in enumerate(ex.stride()):
+            if val in (0, 1):
+                stride[i] = sympy.Integer(val)
+        while any(x is None for x in stride):
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:338]: Quadratic police 🚨

File: torch/fx/experimental/symbolic_shapes.py (hunk 1 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:372]: Sure I'll move it from being a magic method. For Inductor we'll just have a shim for now.

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -329,28 +329,62 @@ def _get_key(self):
...
+        py_sym_int = PySymInt(expr, self)
+        cpp_sym_int = torch.SymIntNode.new_symint(py_sym_int)  # type: ignore[attr-defined]
+        return cpp_sym_int
+
+    def __getitem__(self, val: int) -> "sympy.Expr":
         if not HAS_SYMPY:
             raise RuntimeError("Need sympy installed to create symbolic shapes")
-
-        # TODO: Put 0/1 specialization in guards
-        if val == 0 or val == 1:
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:340]: This is done in the `val_to_var` dict now.

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -329,28 +329,62 @@ def _get_key(self):
...
-            return val
+        if val < 0:
+            # all variables are positive
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:376]: This is only for creating symbolic variables.

File: torch/fx/experimental/symbolic_shapes.py (hunk 2 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:338]: Ehhhh..........

File: functorch/_src/aot_autograd.py [hunk 5]
```diff
@@ -474,19 +475,26 @@ def create_aot_dispatcher_function(
...
         def process_inputs(flat_args):
             if config.use_fake_tensor:
-                def convert(x):
-                    return fake_mode.from_tensor(x, shape_env=shape_env)
-
-                return pytree.tree_map_only(Tensor, convert, flat_args)
+                def convert(idx, x):
+                    if not isinstance(x, torch.Tensor):
+                        return x
+                    if idx < aot_config.num_params_buffers and config.static_weight_shapes:
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:493]: Just checking - is the calling convention here that if `create_aot_dispatcher_function` takes in `n` tensor inputs:
- The first `aot_config.num_params_buffers` tensors correspond to parameters/buffers
- The remaining `n - num_params_buffers` are normal tensor inputs?

Maybe worth mentioning the calling convention in a comment?

File: functorch/_src/config.py [hunk 6]
```diff
@@ -23,3 +23,5 @@
 debug_joint = os.environ.get('AOT_FX_GRAPHS_JOINT', False)
 
 use_dynamic_shapes = os.getenv('AOT_DYNAMIC_SHAPES', False)
+
+static_weight_shapes = True
```
Comment (bdhirsh) [inline functorch/_src/config.py:27]: just curious - what do you think is the end state for how we want to treat parameters <> dynamic shapes - universally on / off / depends on the model? I originally thought the idea was that parameter shapes don't really change in a model, so we always want them to be treated as static, but that sounds wrong now.

File: functorch/test/test_aotdispatch.py [hunk 7]
```diff
@@ -886,6 +886,7 @@ def assert_compiler(gm: torch.fx.GraphModule, _):
     xfail('nn.functional.avg_pool1d', ''),  # Cannot call sizes() on tensor with symbolic sizes/strides
     xfail('nn.functional.avg_pool2d', ''),  # aten.avg_pool2d.default - couldn't find symbolic meta function/...
     xfail('nn.functional.avg_pool3d', ''),  # aten.avg_pool3d.default - couldn't find symbolic meta function/...
+    skip('nn.functional.batch_norm', ''),  # '0 is not tracked with proxy for <torch.fx.experimental.proxy_te..
```
Comment (bdhirsh) [inline functorch/test/test_aotdispatch.py:889]: I'll try to remember to debug / remove this when https://github.com/pytorch/pytorch/pull/84246/commits/247d5db4951f312e6ac86fe59b22df13b68329d4?notification_referrer_id=NT_kwDOAPjlw7M0Mjk0NzkyNjMxOjE2MzExNzQ3 makes it to master

File: torch/fx/experimental/symbolic_shapes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (albanD) [reply torch/fx/experimental/symbolic_shapes.py:374]: Would it be clearer to say "all sympy base variables must be positive" ?

[the remark this answers is not in the document]
Comment (bdhirsh) [reply torch/fx/experimental/symbolic_shapes.py:374]: just curious - is this just to establish some nice invariants? (all "base" values of a sympy expression are positive). Or does something break if we allow negative valued symbols.

[the remark this answers is not in the document]
Comment (bdhirsh) [reply torch/fx/experimental/symbolic_shapes.py:374]: oh, hopefully answered my own question - it looks like in `_maybe_evaluate_static` ([here](https://github.com/pytorch/pytorch/blob/3356d0385fd0f3f0f6ce2d8c681a40fd110c7848/torch/fx/experimental/symbolic_shapes.py#L394)), we can take an expression like `Eq(symbol, 0)` and statically evaluate it to False without introducing guards, since we know that all of our symbols are positive?

File: functorch/_src/aot_autograd.py (hunk 5 above)
Comment (Chillee) [reply functorch/_src/aot_autograd.py:493]: Yeah that's correct, I'll add a note.

File: functorch/_src/config.py (hunk 6 above)
Comment (Chillee) [reply functorch/_src/config.py:27]: ah, good catch, I should just remove the `nn.parameter` check.

**Comment by Chillee** [comment]
@bdhirsh The main unification here for now is simply that Inductor's guards are in terms of the same "symbols" as the ones we have in AOTAutograd.

File: test/test_dynamic_shapes.py [hunk 8]
```diff
@@ -204,7 +205,7 @@ def test_symint_args(self):
         y = create_symbolic_tensor("y", torch.randn(5, 4, 1), shape_env)
         LAST_DIM = 2
         z = x.narrow_copy(LAST_DIM, 0, y.shape[LAST_DIM])
-        self.assertTrue(z.shape[2] == int(y.shape[2]))
```
Comment (wconstab) [inline test/test_dynamic_shapes.py:207]: why this change?

File: test/test_proxy_tensor.py [hunk 9]
```diff
@@ -1066,6 +1066,7 @@ def f(a, b, c, d, e):
     xfail('cholesky_solve', ''),  # Could not run 'aten::_cholesky_solve_helper' with arguments from the 'Meta' back...
     xfail('chunk', ''),  # aten.size.default - couldn't find symbolic meta function/decomposition
     xfail('column_stack', ''),  # Tensors of type TensorImpl do not have numel
+    xfail('combinations', ''),
```
Comment (wconstab) [inline test/test_proxy_tensor.py:1069]: new bug?

File: test/test_proxy_tensor.py [hunk 10]
```diff
@@ -1301,6 +1302,7 @@ def f(a, b, c, d, e):
     xfail('unbind', ''),  # aten.unbind.int - couldn't find symbolic meta function/decomposition
 }
 symbolic_tensor_segfaults = {
+    skip('nn.functional.batch_norm')  # Segfault??
```
Comment (wconstab) [inline test/test_proxy_tensor.py:1305]: hmm.. debug before land?

File: torch/fx/experimental/symbolic_shapes.py [hunk 11]
```diff
@@ -329,28 +329,68 @@ def _get_key(self):
...
 
-    # NB: This is only called for input symbolic sizes; intermediate symbolic
-    # sizes are allocated via a different mechanism
-    def create_symint(self, name, val):
-        assert val >= 0
+    def create_symbolic_sizes_strides(self, ex: torch.Tensor):
+        """
+        Returns a list of symbolic sizes and strides for the given tensor.
+        We try our best to express stride in terms of the sizes, so as to not
+        introduce new symbolic variables.
```
Comment (bdhirsh) [inline torch/fx/experimental/symbolic_shapes.py:336]: Hmm, when would we be in a situation where we actually need to introduce new symbolic variables for any of the strides?

It seems like strides can/should always be expressed in terms of size variables (either from the same tensor, or from a different tensor), but maybe I'm just not thinking of an edge case where this can happen.

File: test/test_dynamic_shapes.py (hunk 8 above)
Comment (Chillee) [reply test/test_dynamic_shapes.py:207]: We used to return integers sometimes, but now we consistently return SymIntNodes (which may wrap integers). I think this is probably better, but open to suggestions.

File: test/test_proxy_tensor.py (hunk 10 above)
Comment (Chillee) [reply test/test_proxy_tensor.py:1305]: I think this is just exposing an underlying bug between functionalization and symbolic shapes. @bdhirsh is looking at it.

Comment (bdhirsh) [reply test/test_proxy_tensor.py:1305]: oh no... what I saw wasn't a segfault 😛. I guess we can see if my patch fixes this though.

Comment (Chillee) [reply test/test_proxy_tensor.py:1305]: It's sometimes a segfault, sometimes the error you reported.

File: torch/fx/experimental/symbolic_shapes.py (hunk 11 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:336]: You can have arbitrarily weird strides.

```
>>> a = torch.arange(1, 20)
>>> unfold_a = a.unfold(0, 5, 2)
>>> unfold_a.stride()
(2, 1)
>>> unfold_a.shape
torch.Size([8, 5])
```

Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:336]: image

Comment (bdhirsh) [reply torch/fx/experimental/symbolic_shapes.py:336]: ah, yep thanks - sometimes I forget how horrible striding logic can be :)

**Review by wconstab — APPROVED** [review]

File: test/test_proxy_tensor.py (hunk 10 above)
Comment (bdhirsh) [reply test/test_proxy_tensor.py:1296]: fwiw, it doesn't look like this has to do with functionalization. I patched my fix, and what I noticed from some poking around is:

(1) The segfault happens [here](https://github.com/pytorch/pytorch/blob/317eeb81c3e7ab21ca4359819c4a89122ce574f5/torch/csrc/jit/python/init.cpp#L1390). At some point, we must be calling `.get_pyobj()` on a symint that isn't actually a `PySymIntNodeImpl` (or just that's in an invalid state).
(2) It happens before any functionalization code runs
